src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
I think it is safer to put `volatile ` for the variables which is only *read* within a loop of progress polling which might be fully inlined without any function calls inside, because the compiler or hardware can optimize the reading. The variables might be updated by the other thread. The function pointer triggered the updating operation can be the callback in progress engine, or potentially the event registered to network driver, thus the compiler may not be able to detect it. And the progress polling macro can be changed in future.

I removed two unneeded `volatile` (the count in `MPIDI_CH4U_win_sync_lock_t` and `MPIDI_CH4U_win_lock_recvd_t `), since they do not work as the above mode.

I also added `volatile` for the completion counters in the last commit. I have explained the reason in code comment for each `volatile`.